src/state_machine.zig

```diff
@@ -42,6 +42,18 @@ pub fn StateMachineType(
         pub const constants = struct {
             pub const message_body_size_max = config.message_body_size_max;
 
+            /// The maximum amount of logical batches that may be queued on a client
+            /// (spread across all requests).
+            pub const batch_logical_max = blk: {
+                var max: usize = 0;
+                inline for (std.enums.values(Operation)) |operation| {
+                    if (@intFromEnum(operation) < config.vsr_operations_reserved) continue;
+                    if (!batch_logical_allowed(operation)) continue;
+                    max = @max(max, batch_max.operation_batch_max(operation));
+                }
+                break :blk max * global_constants.client_request_queue_max;
+            };
+
```

Comment:
> That is an improvement, but why not go with a context-style API:
>
> - Then we can get rid of batch_logical_max and client_request_queue_max? (The limit moves up a layer. Clients could have no limit (by using dynamic allocator), or their own static limit.)
> - Reduced Client surface area: Back to a single request function, instead of both batch_get and batch_submit.
> - Reduced Client surface area: I think there would be no possible error on the request method.

I agree we can improve the client API, however, I thought it was out of the scope of this PR.
Can you elaborate on the suggestion of a context-style API a bit further?

...

The current implementation of the `tb_client` can take advantage of 4096 batches at most since it's the hard-coded upper `concurrency_max` limit.
That's why I think it makes sense for the caller to save memory specifying a new limit, lower than `batch_logical_max`.

Reply:
> I agree we can improve the client API, however, I thought it was out of the scope of this PR.

Sounds good to me.